### PR TITLE
Renames to camelCase, allow map to accept a few more syntaxes and fix find.

### PR DESCRIPTION
**Pull request by Rekihyt**
I realize this should probably be more than one pr, let me know if I should split it up.

### CODE REVIEW

File: zero_functional.nim [hunk 1]
```diff
@@ -39,8 +42,8 @@ else:
     const defaultCollectionType = "seq"
   const defaultResultType = defaultCollectionType & "[int]"
 
-proc print_code(code: NimNode) =
-  echo(repr(code).replace("`gensym", "_").replace("__call__", "Call").replace("__", ""))
+proc printCode(code: NimNode) =
+  echo(repr(code).replace("`gensym", "_")#[.replace(callSuffix, "Call")]#.replace("_", ""))
```
Comment (michael72) [inline zero_functional.nim:46]: this should be restored - see comment above concerning the internal variable names

File: zero_functional.nim [hunk 2]
```diff
@@ -1243,16 +1258,20 @@ zf_inline exists(search: bool):
...
-    result = none(it.type)
+    var initAccu = true
   loop:
-    if cond:
+    if initAccu:
+      # type of it must be assigned here because it may have been changed
+      result = none(it.type)
+      initAccu = false
+    elif cond:
       return some(it)
```
Comment (michael72) [inline zero_functional.nim:1270]: why was this change necessary?

File: zero_functional.nim [hunk 3]
```diff
@@ -1415,7 +1434,7 @@ proc combineWithOtherCollection(ext: ExtNimNode, indexed: bool) {.compileTime.}
     let listRef = ext.node[idx]
     idx += 1
     if indexed:
-      var idxInner = genSym(nskVar, "__idxInner__")
+      var idxInner = genSym(nskVar, "_idxInner")
```
Comment (michael72) [inline zero_functional.nim:1437]: see comment above - change back

File: test.nim [hunk 4]
```diff
@@ -1030,6 +1049,12 @@ suite "valid chains":
     check(p.yes == @[2, 4, 6])
     check(p.no == @[1, 3, 5])
 
+  # test "partition after changed type":
+  #   let a = @[111, 2, 33, 4, 5555, 6]
+  #   let p = a --> map($it) --> partition(it.len() == 1)
+  #   check(p.yes == @["2", "4", "6"])
+  #   check(p.no == @["111", "33", "5555"])
+
```
[the remark this answers is not in the document]
Comment (michael72) [reply test.nim:1057]: unless you forget to remove the comments - there should not be any commented-out code sections

[the remark this answers is not in the document]
Comment (michael72) [reply test.nim:1057]: ... ok I see the problem now - seems like there is a problem with the `init` section when in a nested loop. I would like to fix that more generally - and not specfically for find, uniq, etc.
Maybe you can leave the commented tests as they are - and I will try to do a general fix then.

[the remark this answers is not in the document]
Comment (Rekihyt) [reply test.nim:1057]: That would be great! The problem arises from type(it) inside init always returning the original element type even if it was changed by something like map.

File: zero_functional.nim (hunk 2 above)
Comment (Rekihyt) [reply zero_functional.nim:1270]: This is a stopgap fix for it.type returning the original element type instead of the current one. It makes test "find with changed type" pass.

**Comment by Rekihyt** [comment]
Please check if I got the enumerate command documentation correct.

**Comment by Rekihyt** [comment]
Well this is slightly embarrassing, didn't realize the --styleCheck:hint I had forgotten in my global config was causing a bunch of name warnings. Ill remove the rest of the name changes unless you still want them.

File: zero_functional.nim (hunk 2 above)
Comment (michael72) [reply zero_functional.nim:1270]: OK - I wrote that comment before I see why the other tests were commented out. Same issue with the `init` section here I guess.
First I want to fix it more generally, then it should still be efficient... Finding out the output type is really tough - that's why this whole thing with the extra proc and using its return type... I tried a lot of stuff but a hit a lot of dead ends there. Maybe meanwhile it is easier in nim...?
If you have any ideas you're most welcome! But I think I have to think about it thoroughly.
You can get ideas of course if you look at the output of `->>` and try to come up with a general solution replacing the generated code by some other new code. Biggest problem when creating the macros is often that some information is retrieved later in the process of parsing - so finding out a return type before its use is tough.
Hope we can find a solution soon :-)

Comment (michael72) [reply zero_functional.nim:1270]: btw - did you check this solution works also when finding the first (or only) element?

**Comment by michael72** [comment]
Hey @Rekihyt 
I think the simple but effective tests for changing the type and applying other functions are great.
I took over almost all of your changes except a few:
- internal naming with dunder __ -> this is easier to remove when printing the syntax tree
- calling conventions with map - I only left the automatic apply of a function but removed `it.f()` because it should be equivalent to `f(it)` in nim. I don't know what `f it` was for - but we surely do not want something like `a.map(+1)` - not sure if I want that. I'd rather go with adding the `_` as an alternative to using `it` - like so: `a.map(_ + 1)` - is a bit like scala semantics.

The zf code now got a bit more complicated - I hope I can find a way to boil things down a little again.

Thanks for your input!

**Comment by michael72** [comment]
@Rekihyt 
If you find other failing tests that would be great of course :-)
